notion-core/src/main/scala/com/ambiata/notion/core/SequenceFileIO.scala

```diff
@@ -0,0 +1,118 @@
+package com.ambiata.notion.core
+
+import com.ambiata.mundane.control._
+import com.ambiata.mundane.io._
+import com.ambiata.poacher.hdfs._
+import com.ambiata.saws.s3._
+import com.ambiata.com.amazonaws.services.s3.AmazonS3Client
+
+import org.apache.hadoop.io.compress.CompressionCodec
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.io.{Writable, BytesWritable, NullWritable, IntWritable, SequenceFile}
+import org.apache.hadoop.fs.{Hdfs => _, _}
+
+import java.io._
+
+import scalaz._, Scalaz._, effect.Effect._
+
+/**
+ * This is based off SequenceUtil, but can read/write sequence files of any key/value type
+ * 
+ * WARNING: Not suitable for reading/writing large amounts of data as everything is kept in memory
+ */
+case class SequenceFileIO[K <: Writable, V <: Writable, A](conf: Configuration, emptyKey: K, emptyValue: V, fromKeyValue: (K, V) => A, toKeyValue: A => (K, V))
+                                                          (implicit km: Manifest[K], vm: Manifest[V]) {
+
+  def readLocation(location: Location, client: AmazonS3Client): RIO[List[A]] = for {
+    in <- (location match {
+      case HdfsLocation(p) =>
+        RIO.safe[InputStream](FileSystem.get(conf).open(new Path(p)))
+      case LocalLocation(p) =>
+        RIO.safe[InputStream](new LocalInputStream(p))
+      case S3Location(bucket, key) =>
+        S3InputStream.stream(S3Address(bucket, key), client)
+    })
+    as <- readInputStream(in)
+  } yield as
+
+  def readInputStream(input: InputStream): RIO[List[A]] =
```

Comment:
Please can e not add a call back, it could definitely be more general, but everytime we add one of these callbacks it takes weeks to unwind it, it is almost never the right thing to do.